### ISSUE TITLE
Looked all papers, none of them really simple yet effective and opensource

### ISSUE THREAD

**Issue by MonolithFoundation** [JinXins/Awesome-Token-Merge-for-MLLMs#1]
I am not sure why there is no such work that is simple (without hacking the layers of large language models), effective (without brute force usage of a single token), and open source.

**Comment by JinXins** [comment]
Yes, most of the methods in these papers are similar. I guess it might be because they were all conducted around the same period.

**Comment by MonolithFoundation** [comment]
In terms of effective and simplicity, even needed retraining, what's the current best method so far

**Comment by JinXins** [comment]
In my opinion, the better method should be sufficient in two ways: to reduce and mine the training data and to distillate to a small model.

**Comment by JinXins** [comment]
In my opinion, the better method should be sufficient in two ways: to reduce and mine the training data and to distillate to a small model.

**Comment by MonolithFoundation** [comment]
Numerous methods utilize Resampler and borrow techniques from various sources. Nevertheless, no one has truly implemented this straightforward approach:
1. Obtaining dynamic tokens using native input.
2. Employing a simple yet potent softmax router on visual tokens to obtain the top k.

Is there any paper that has explored this approach in depth?

**Comment by JinXins** [comment]
I'm not sure there are have paper utilize these ways you metioned. As you can see, I have had some days not updating this repo since I have struggling my PhD application. If you seen any papers related it, please tell me.

**Comment by MonolithFoundation** [comment]
Yes, DeepSeek did it, he gives me a module, seems can work.

**Comment by JinXins** [comment]
That's great!

**Comment by MonolithFoundation** [comment]
Can u give me some advise on his solution?

```python

Copy
import torch
import torch.nn as nn

class TopKSelector(nn.Module):
    def __init__(self, feature_dim=1024, k=10):
        super().__init__()
        self.k = k
        # 可学习的查询向量
        self.q = nn.Parameter(torch.randn(feature_dim))
        # 可选：归一化
        self.norm = nn.LayerNorm(feature_dim)
        
    def forward(self, vit_features):
        # vit_features: [B, T, 1024]
        B, T, _ = vit_features.shape
        
        # 归一化特征和查询向量（可选）
        features_norm = self.norm(vit_features)
        q_norm = nn.functional.normalize(self.q, dim=0)
        
        # 计算分数 [B, T]
        scores = torch.matmul(features_norm, q_norm)
        
        # 取TopK索引 [B, K]
        _, indices = torch.topk(scores, self.k, dim=1)
        
        # 收集特征 [B, K, 1024]
        selected_tokens = torch.gather(
            vit_features, 
            dim=1, 
            index=indices.unsqueeze(-1).expand(-1, -1, vit_features.size(-1))
        
        return selected_tokens
```

**Comment by jjihwan** [comment]
Hi @MonolithFoundation 
seems very interesting discussion. Where can I find this code? DeepSeek-VL? or other model?

> Can u give me some advise on his solution?
> 
> Copy
> import torch
> import torch.nn as nn
> 
> class TopKSelector(nn.Module):
>     def __init__(self, feature_dim=1024, k=10):
>         super().__init__()
>         self.k = k
>         # 可学习的查询向量
>         self.q = nn.Parameter(torch.randn(feature_dim))
>         # 可选：归一化
>         self.norm = nn.LayerNorm(feature_dim)
>         
>     def forward(self, vit_features):
>         # vit_features: [B, T, 1024]
>         B, T, _ = vit_features.shape
>         
>         # 归一化特征和查询向量（可选）
>         features_norm = self.norm(vit_features)
>         q_norm = nn.functional.normalize(self.q, dim=0)
>         
>         # 计算分数 [B, T]
>         scores = torch.matmul(features_norm, q_norm)
>         
>         # 取TopK索引 [B, K]
>         _, indices = torch.topk(scores, self.k, dim=1)
>         
>         # 收集特征 [B, K, 1024]
>         selected_tokens = torch.gather(
>             vit_features, 
>             dim=1, 
>             index=indices.unsqueeze(-1).expand(-1, -1, vit_features.size(-1))
>         
>         return selected_tokens
